src/BSON/Value.c

```diff
@@ -0,0 +1,631 @@
+}
+
+static PHP_METHOD(MongoDB_BSON_Value, getBool)
+{
+	php_phongo_value_t* intern;
+
+	PHONGO_PARSE_PARAMETERS_NONE();
+
+	intern = Z_VALUE_OBJ_P(getThis());
+
+	PHONGO_VALUE_CHECK_TYPE(intern->value, BSON_TYPE_BOOL);
+
+	RETURN_BOOL(intern->value.value.v_bool);
+}
+
+static PHP_METHOD(MongoDB_BSON_Value, getCode)
+{
+	php_phongo_value_t* intern;
+	bson_t*             scope;
+
+	PHONGO_PARSE_PARAMETERS_NONE();
+
+	intern = Z_VALUE_OBJ_P(getThis());
+
+	switch (intern->value.value_type) {
+		case BSON_TYPE_CODE:
+			php_phongo_bson_new_javascript(return_value, intern->value.value.v_code.code, intern->value.value.v_code.code_len, NULL);
+			return;
+
+		case BSON_TYPE_CODEWSCOPE:
+			scope = bson_new_from_data(intern->value.value.v_codewscope.scope_data, intern->value.value.v_codewscope.scope_len);
+			php_phongo_bson_new_javascript(return_value, intern->value.value.v_codewscope.code, intern->value.value.v_codewscope.code_len, scope);
+			return;
+
+		default:
+			phongo_throw_exception(PHONGO_ERROR_UNEXPECTED_VALUE, "Unexpected BSON type %d", intern->value.value_type);
+	}
+}
+
+static PHP_METHOD(MongoDB_BSON_Value, getDBPointer)
+{
+	php_phongo_value_t* intern;
+
+	PHONGO_PARSE_PARAMETERS_NONE();
+
+	intern = Z_VALUE_OBJ_P(getThis());
+
+	PHONGO_VALUE_CHECK_TYPE(intern->value, BSON_TYPE_DBPOINTER);
+
+	php_phongo_bson_new_dbpointer(return_value, intern->value.value.v_dbpointer.collection, intern->value.value.v_dbpointer.collection_len, &intern->value.value.v_dbpointer.oid);
+}
+
+static PHP_METHOD(MongoDB_BSON_Value, getDecimal128)
+{
+	php_phongo_value_t* intern;
+
+	PHONGO_PARSE_PARAMETERS_NONE();
+
+	intern = Z_VALUE_OBJ_P(getThis());
+
+	PHONGO_VALUE_CHECK_TYPE(intern->value, BSON_TYPE_DECIMAL128);
+
+	php_phongo_bson_new_decimal128(return_value, &intern->value.value.v_decimal128);
+}
+
+static PHP_METHOD(MongoDB_BSON_Value, getDocument)
+{
+	php_phongo_value_t* intern;
+	bson_t*             bson;
+
+	PHONGO_PARSE_PARAMETERS_NONE();
+
+	intern = Z_VALUE_OBJ_P(getThis());
+
+	PHONGO_VALUE_CHECK_TYPE(intern->value, BSON_TYPE_DOCUMENT);
+
+	bson = bson_new_from_data(intern->value.value.v_doc.data, intern->value.value.v_doc.data_len);
+
+	php_phongo_bson_new_document(return_value, bson, false);
+}
+
+static PHP_METHOD(MongoDB_BSON_Value, getDouble)
+{
+	php_phongo_value_t* intern;
+
+	PHONGO_PARSE_PARAMETERS_NONE();
+
+	intern = Z_VALUE_OBJ_P(getThis());
+
+	PHONGO_VALUE_CHECK_TYPE(intern->value, BSON_TYPE_DOUBLE);
+
+	RETURN_DOUBLE(intern->value.value.v_double);
+}
+
+static PHP_METHOD(MongoDB_BSON_Value, getInt)
+{
+	php_phongo_value_t* intern;
+
+	PHONGO_PARSE_PARAMETERS_NONE();
+
+	intern = Z_VALUE_OBJ_P(getThis());
+
+	switch (intern->value.value_type) {
+		case BSON_TYPE_INT32:
+			RETURN_LONG(intern->value.value.v_int32);
+
+		case BSON_TYPE_INT64:
+			RETURN_INT64(intern->value.value.v_int64);
```

Comment:
This will always an `int` on 64-bit platforms and only use an `Int64` instance on 32-bit platforms if the value exceeds the 64-bit range. For the time being I consider this the better way, as `$value->getType()` and `$value->isInt64()` are available to check if the data received was a 64-bit integer.